src/transformers/trainer.py

```diff
@@ -3464,15 +3487,52 @@ def _save_tpu(self, output_dir: Optional[str] = None):
         model = self.model
         xm.mark_step()
 
-        if xm.is_master_ordinal():
+        if xm.is_master_ordinal(local=False):
             os.makedirs(output_dir, exist_ok=True)
             torch.save(self.args, os.path.join(output_dir, TRAINING_ARGS_NAME))
 
         # Save a trained model and configuration using `save_pretrained()`.
         # They can then be reloaded using `from_pretrained()`
         supported_classes = (PushToHubMixin,)
         xm.rendezvous("saving_checkpoint")
-        if not isinstance(model, supported_classes):
+        if self.is_fsdp_xla_enabled and not self.is_fsdp_xla_v2_enabled:
+            ckpt = {
+                "model": model.state_dict(),
+                "shard_metadata": model.get_shard_metadata(),
+            }
+            ckpt_path = os.path.join(
+                output_dir, f"rank{self.args.process_index}_of_{self.args.world_size}_{WEIGHTS_NAME}.pth"
+            )
+            # All ranks save sharded checkpoint
+            xm.save(ckpt, ckpt_path, master_only=False)
+            # Make sure all ranks have saved checkpoints
+            xm.rendezvous("save_full_checkpoints")
+            # Master save full checkpoint
+            if self.args.should_save:
+                from torch_xla.distributed.fsdp import XlaFullyShardedDataParallel as FSDP
+                from torch_xla.distributed.fsdp import consolidate_sharded_model_checkpoints
+
+                full_state_dict, _ = consolidate_sharded_model_checkpoints(
+                    ckpt_prefix=os.path.join(output_dir, ""),
+                    ckpt_suffix=f"rank*_of_*_{WEIGHTS_NAME}.pth",
+                    save_model=False,
+                )
+                assert isinstance(model, FSDP)
```

Comment:
We don't do asserts in the codebase. If we worry about if this is not a thing, we should do an `if not` + raise a proper and clear `error`

Reply:
OK, I remove this check since the model should be FSDP XLA when `self.is_fsdp_xla_enabled and not self.is_fsdp_xla_v2_enabled`.